persist/sqlite/accounts.go

```diff
@@ -7,12 +7,128 @@ import (
+		}
+
+		total := usage.RenterCost()
+		balance, underflow := balance.SubWithUnderflow(total)
+		if underflow {
+			return fmt.Errorf("insufficient balance")
+		}
+
+		_, err = tx.Exec(`UPDATE accounts SET balance=$1 WHERE id=$2`, encode(balance), dbID)
+		if err != nil {
+			return fmt.Errorf("failed to update balance: %w", err)
+		} else if err := updateV2ContractFunding(tx, dbID, usage); err != nil {
+			return fmt.Errorf("failed to update contract funding: %w", err)
+		}
+		return nil
+	})
+}
+
+// RHP4CreditAccounts credits the accounts with the given deposits and revises
+// the contract.
+func (s *Store) RHP4CreditAccounts(deposits []proto4.AccountDeposit, contractID types.FileContractID, revision types.V2FileContract) (balances []types.Currency, err error) {
+	err = s.transaction(func(tx *txn) error {
+		getBalanceStmt, err := tx.Prepare(`SELECT balance FROM accounts WHERE account_id=$1`)
+		if err != nil {
+			return fmt.Errorf("failed to prepare get balance statement: %w", err)
+		}
+		defer getBalanceStmt.Close()
+
+		updateBalanceStmt, err := tx.Prepare(`INSERT INTO accounts (account_id, balance, expiration_timestamp) VALUES ($1, $2, $3) ON CONFLICT (account_id) DO UPDATE SET balance=EXCLUDED.balance, expiration_timestamp=EXCLUDED.expiration_timestamp RETURNING id`)
+		if err != nil {
+			return fmt.Errorf("failed to prepare update balance statement: %w", err)
+		}
+		defer updateBalanceStmt.Close()
+
+		getFundingAmountStmt, err := tx.Prepare(`SELECT amount FROM contract_v2_account_funding WHERE contract_id=$1 AND account_id=$2`)
+		if err != nil {
+			return fmt.Errorf("failed to prepare get funding amount statement: %w", err)
+		}
+		defer getFundingAmountStmt.Close()
+
+		updateFundingAmountStmt, err := tx.Prepare(`INSERT INTO contract_v2_account_funding (contract_id, account_id, amount) VALUES ($1, $2, $3) ON CONFLICT (contract_id, account_id) DO UPDATE SET amount=EXCLUDED.amount`)
+		if err != nil {
+			return fmt.Errorf("failed to prepare update funding amount statement: %w", err)
+		}
+		defer updateFundingAmountStmt.Close()
+
+		var contractDBID int64
+		err = tx.QueryRow(`SELECT id FROM contracts_v2 WHERE contract_id=$1`, encode(contractID)).Scan(&contractDBID)
+		if err != nil {
+			return fmt.Errorf("failed to get contract ID: %w", err)
+		}
+
+		var usage proto4.Usage
+		var createdAccounts int
+		for _, deposit := range deposits {
+			var balance types.Currency
+			err := getBalanceStmt.QueryRow(encode(deposit.Account)).Scan(decode(&balance))
+			if err != nil && !errors.Is(err, sql.ErrNoRows) {
+				return fmt.Errorf("failed to get balance: %w", err)
+			} else if errors.Is(err, sql.ErrNoRows) {
+				createdAccounts++
+			}
+
+			balance = balance.Add(deposit.Amount)
+
+			var accountDBID int64
+			err = updateBalanceStmt.QueryRow(encode(deposit.Account), encode(balance), encode(time.Now().Add(90*24*time.Hour))).Scan(&accountDBID)
+			if err != nil {
+				return fmt.Errorf("failed to update balance: %w", err)
```

Comment:
`nit` the account ID might be useful here

Reply:
account ID doesn't matter if the query fails. It's unlikely to be related to the account ID.